File: packages/provider/src/responses.ts
/**
 * The OpenAI Responses API stream, as Demi reads it. Shared by every adapter
 * that speaks it (the Codex backend and the OpenAI Responses transport).
 *
 * These schemas describe only what Demi consumes; `looseObject` keeps every
 * other field the vendor sends, so a decoded item can be re-sent or stored
 * whole. Unknown event and item types decode to `null` and are ignored, while
 * a known type with a malformed payload is a protocol error naming the field.
 */
import { z } from 'zod'
import { tokenUsageWithCachedInput } from './usage'
import {
  reportedStringSchema,
  taggedUnion,
  tokenCountSchema
} from './vendor-schema'
import type { TokenUsage } from '@demicodes/core'

/** A vendor error object, as `response.failed` and `error` events carry it. */
export const responsesErrorSchema = z.looseObject({
  code: reportedStringSchema,
  type: reportedStringSchema,
  message: reportedStringSchema,
  request_id: reportedStringSchema,
  requestId: reportedStringSchema,
})
export type ResponsesError = z.infer<typeof responsesErrorSchema>

export const responsesUsageSchema = z.looseObject({
  input_tokens: tokenCountSchema,
  output_tokens: tokenCountSchema,
  input_tokens_details: z.looseObject({ cached_tokens: tokenCountSchema })
    .optional(),
})
export type ResponsesUsage = z.infer<typeof responsesUsageSchema>

/** The token usage a `response.completed` event reports. */
export function tokenUsageFromResponsesUsage(
  usage: ResponsesUsage | null | undefined
): TokenUsage {
  return tokenUsageWithCachedInput({
    inputTokens: usage?.input_tokens,
    outputTokens: usage?.output_tokens,
    cachedTokens: usage?.input_tokens_details?.cached_tokens,
  })
}

/** One part of a reasoning item's `summary` or `content`. */
const reasoningPartSchema = z.looseObject({
  type: z.string().optional(),
  text: z.string(),
})

/** One part of a message item's `content`; unknown parts decode to null. */
export const responsesMessagePartSchema = taggedUnion({
  output_text: z.looseObject({
    type: z.literal('output_text'),
    text: z.string(),
  }),
  refusal: z.looseObject({
    type: z.literal('refusal'),
    refusal: z.string(),
  }),
})
export type ResponsesMessagePart =
  NonNullable<z.infer<typeof responsesMessagePartSchema>>

export const responsesReasoningItemSchema = z.looseObject({
  type: z.literal('reasoning'),
  id: z.string().optional(),
  summary: z.array(reasoningPartSchema).optional(),
  content: z.array(reasoningPartSchema).optional(),
  encrypted_content: z.string().optional(),
})
export type ResponsesReasoningItem =
  z.infer<typeof responsesReasoningItemSchema>

export const responsesMessageItemSchema = z.looseObject({
  type: z.literal('message'),
  id: z.string().optional(),
  role: z.string().optional(),
  content: z.array(responsesMessagePartSchema).optional(),
  status: z.string().optional(),
})
export type ResponsesMessageItem = z.infer<typeof responsesMessageItemSchema>

export const responsesFunctionCallItemSchema = z.looseObject({
  type: z.literal('function_call'),
  id: z.string().optional(),
  call_id: z.string().optional(),
  name: z.string().optional(),
  arguments: z.string().optional(),
})
export type ResponsesFunctionCallItem =
  z.infer<typeof responsesFunctionCallItemSchema>

/** An output item of the three kinds Demi maps; any other decodes to null. */
export const responsesItemSchema = taggedUnion({
  reasoning: responsesReasoningItemSchema,
  message: responsesMessageItemSchema,
  function_call: responsesFunctionCallItemSchema,
})
export type ResponsesItem = NonNullable<z.infer<typeof responsesItemSchema>>

/** A `…delta` event: one increment of the text the vendor is streaming. */
function deltaEvent<T extends string>(type: T) {
  return z.looseObject({
    type: z.literal(type),
    delta: z.string(),
  })
}

/** The stream events Demi maps; any other event type decodes to null. */
export const responsesEventSchema = taggedUnion({
  'response.output_item.added': z.looseObject({
    type: z.literal('response.output_item.added'),
    item: responsesItemSchema.optional(),
  }),
  'response.output_item.done': z.looseObject({
    type: z.literal('response.output_item.done'),
    item: responsesItemSchema.optional(),
    item_id: z.string().optional(),
    call_id: z.string().optional(),
  }),
  'response.output_text.delta': deltaEvent('response.output_text.delta'),
  'response.reasoning_text.delta': deltaEvent('response.reasoning_text.delta'),
  'response.reasoning_summary_text.delta':
    deltaEvent('response.reasoning_summary_text.delta'),
  'response.function_call_arguments.delta': z.looseObject({
    type: z.literal('response.function_call_arguments.delta'),
    item_id: z.string().optional(),
    delta: z.string(),
  }),
  'response.function_call_arguments.done': z.looseObject({
    type: z.literal('response.function_call_arguments.done'),
    item_id: z.string().optional(),
    arguments: z.string(),
  }),
  'response.completed': z.looseObject({
    type: z.literal('response.completed'),
    response: z.looseObject({ usage: responsesUsageSchema.optional() })
      .optional(),
  }),
  'response.failed': z.looseObject({
    type: z.literal('response.failed'),
    response: z.looseObject({
      id: reportedStringSchema,
      error: responsesErrorSchema.optional(),
    }).optional(),
  }),
  'response.incomplete': z.looseObject({
    type: z.literal('response.incomplete'),
    response: z.looseObject({
      incomplete_details: z.looseObject({ reason: reportedStringSchema })
        .optional(),
    }).optional(),
  }),
  // The SSE stream sends {type:'error', message, code}; the Codex WebSocket
  // backend nests the same failure as {type:'error', error:{…}}.
  error: z.looseObject({
    type: z.literal('error'),
    message: reportedStringSchema,
    code: reportedStringSchema,
    error: responsesErrorSchema.optional(),
  }),
})
export type ResponsesEvent = NonNullable<z.infer<typeof responsesEventSchema>>

/**
 * Decodes one Responses stream event. Returns null for an event type Demi does
 * not map; throws when a mapped type arrives malformed.
 */
export function decodeResponsesEvent(raw: unknown): ResponsesEvent | null {
  return responsesEventSchema.parse(raw)
}

File: crates/provider/tests/provider/responses.rs
//! The Responses stream mapped onto a run's events (`providers.md` § A run,
//! § Reading vendor input), over recorded frames.

use std::{sync::Arc, time::Duration};

use bytes::Bytes;
use demi_core::{FailureSource, ProviderErrorDiagnostics, ProviderFailureFacts, Timestamp, TokenUsage};
use demi_provider::{
    ErrorCode, ProviderEvent, ProviderFailure, ToolCall, read_http_failure,
    testing::FixedClock,
    wire::{
        Vendor,
        responses::{decode_frame, map_events, sse_events},
    },
};
use futures_util::{StreamExt, stream};
use serde_json::{Value, json};
use tokio_util::sync::CancellationToken;

const NOW: &str = "2026-09-18T14:00:00.000Z";

fn vendor(reader: demi_provider::FailureReader) -> Vendor {
    Vendor {
        label: "Codex",
        reader,
        clock: Arc::new(FixedClock(NOW.parse::<Timestamp>().unwrap())),
    }
}

/// A body with one `data:` frame per payload; a string is sent as it is.
fn body(frames: &[Value]) -> impl futures_util::Stream<Item = reqwest::Result<Bytes>> + 'static {
    let text: String = frames
        .iter()
        .map(|frame| match frame {
            Value::String(text) => format!("data: {text}\n\n"),
            other => format!("data: {other}\n\n"),
        })
        .collect();
    stream::iter([Ok(Bytes::from(text))])
}

async fn events_with(frames: &[Value], reader: demi_provider::FailureReader) -> Vec<ProviderEvent> {
    let cancel = CancellationToken::new();
    let events = sse_events(body(frames), "Codex", cancel.clone());
    map_events(events, vendor(reader), "codex:", cancel).collect().await
}

async fn events(frames: &[Value]) -> Vec<ProviderEvent> {
    events_with(frames, read_http_failure).await
}

fn failure_of(events: Vec<ProviderEvent>) -> ProviderFailure {
    match <[ProviderEvent; 1]>::try_from(events) {
        Ok([ProviderEvent::Error(failure)]) => failure,
        other => panic!("{other:?}"),
    }
}

#[tokio::test]
async fn a_stream_maps_thinking_text_tool_calls_and_usage() {
    let reasoning = json!({
        "type": "reasoning", "id": "rs_1", "encrypted_content": "enc",
        "summary": [{ "type": "summary_text", "text": "thought" }], "status": "completed",
    });
    let events = events(&[
        json!({ "type": "response.created", "response": { "id": "resp-1" } }),
        json!({ "type": "response.output_item.added", "item": { "type": "reasoning", "id": "rs_1" } }),
        json!({ "type": "response.reasoning_summary_text.delta", "delta": "think", "sequence_number": 3 }),
        json!({ "type": "response.output_item.done", "item": reasoning }),
        json!({ "type": "response.output_text.delta", "delta": "hello " }),
        json!({ "type": "response.output_item.added", "item": { "type": "function_call", "id": "fc_1", "call_id": "call_1", "name": "shell_exec", "arguments": "" } }),
        json!({ "type": "response.function_call_arguments.delta", "item_id": "fc_1", "delta": "{\"script\":" }),
        json!({ "type": "response.function_call_arguments.done", "item_id": "fc_1", "arguments": "{\"script\":\"pwd\"}" }),
        json!({ "type": "response.output_item.done", "item": { "type": "function_call", "id": "fc_1", "call_id": "call_1", "name": "shell_exec" } }),
        json!({ "type": "response.completed", "response": { "usage": { "input_tokens": 100, "output_tokens": 7, "input_tokens_details": { "cached_tokens": 60 } } } }),
        json!("[DONE]"),
    ])
    .await;
    let [start, delta, ProviderEvent::ThinkingSignature(signature), text, call, response] = events.as_slice() else {
        panic!("{events:?}");
    };
    assert_eq!((start, delta, text), (&ProviderEvent::ThinkingStart, &ProviderEvent::ThinkingDelta("think".into()), &ProviderEvent::TextDelta("hello ".into())));
    // The signature is the whole item, the fields Demi does not read
    // included, tagged as this provider's.
    let item: Value = serde_json::from_str(signature.strip_prefix("codex:").unwrap()).unwrap();
    assert_eq!(item, reasoning);
    assert_eq!(
        call,
        &ProviderEvent::ToolCall(ToolCall {
            tool_use_id: "call_1|fc_1".into(),
            tool_name: "shell_exec".into(),
            input: json!({ "script": "pwd" }),
        })
    );
    let usage = TokenUsage {
        input_tokens: 40,
        output_tokens: 7,
        cache_read_tokens: 60,
        cache_write_tokens: 0,
    };
    assert_eq!(response, &ProviderEvent::Response(usage));
}

#[tokio::test]
async fn a_finished_items_text_is_emitted_only_when_no_delta_streamed_it() {
    let message = |text: &str| json!({ "type": "response.output_item.done", "item": { "type": "message", "role": "assistant", "content": [{ "type": "output_text", "text": text, "annotations": [] }, { "type": "output_audio", "data": "…" }] } });
    let events = events(&[
        json!({ "type": "response.output_item.added", "item": { "type": "reasoning", "id": "rs_1" } }),
        json!({ "type": "response.output_item.done", "item": { "type": "reasoning", "id": "rs_1", "content": [{ "text": "raw reasoning" }], "encrypted_content": "enc" } }),
        json!({ "type": "response.output_text.delta", "delta": "因为" }),
        json!({ "type": "response.output_text.delta", "delta": "天空是蓝的" }),
        message("因为天空是蓝的"),
        message("and the next message"),
        json!({ "type": "response.output_item.done", "item": { "type": "web_search_call", "status": "completed" } }),
        json!({ "type": "response.output_item.done", "item": { "type": "message", "content": [{ "type": "refusal", "refusal": "no" }] } }),
        json!({ "type": "response.completed" }),
    ])
    .await;
    let texts: Vec<&str> = events
        .iter()
        .filter_map(|event| match event {
            ProviderEvent::TextDelta(text) | ProviderEvent::ThinkingDelta(text) => Some(text.as_str()),
            _ => None,
        })
        .collect();
    assert_eq!(texts, ["raw reasoning", "因为", "天空是蓝的", "and the next message", "no"]);
    assert_eq!(events.last(), Some(&ProviderEvent::Response(TokenUsage::default())));
}

#[tokio::test]
async fn a_stream_that_ends_without_its_completion_still_responds_with_zero_usage() {
    let events = events(&[json!({ "type": "response.output_text.delta", "delta": "hi" })]).await;
    assert_eq!(events, [ProviderEvent::TextDelta("hi".into()), ProviderEvent::Response(TokenUsage::default())]);
}

#[tokio::test]
async fn a_failure_keeps_the_frame_as_its_record_and_takes_the_wait_from_the_providers_reader() {
    // A field in a form Demi does not read survives: the record is the text
    // as it came.
    let frame = r#"{"type":"error","error":{"type":"usage_limit_reached","message":"The usage limit has been reached","resets_at":"soon"},"status_code":"429"}"#;
    let in_a_minute: demi_provider::FailureReader = |diagnostics, received_at| {
        assert_eq!(diagnostics.upstream.as_deref(), Some(r#"{"type":"error","error":{"type":"usage_limit_reached","message":"The usage limit has been reached","resets_at":"soon"},"status_code":"429"}"#));
        let later = Timestamp::from_millisecond(received_at.as_millisecond() + 60_000).unwrap();
        ProviderFailureFacts { retry_at: Some(later) }
    };
    let failure = failure_of(events_with(&[Value::String(frame.into())], in_a_minute).await);
    assert_eq!(failure.message, "The usage limit has been reached");
    assert_eq!(failure.code, Some(ErrorCode::RateLimit));
    assert_eq!(failure.retry_after, Some(Duration::from_secs(60)));
    let diagnostics = failure.diagnostics.unwrap();
    assert_eq!((diagnostics.source, diagnostics.provider_code.as_deref()), (FailureSource::Stream, Some("usage_limit_reached")));
    assert_eq!(diagnostics.upstream.as_deref(), Some(frame));
}

#[tokio::test]
async fn failed_incomplete_and_error_events_end_the_run_classified_and_naming_the_vendor() {
    let cases = [
        (json!({ "type": "response.failed", "response": { "error": { "code": "context_length_exceeded", "message": "too long" } } }), "too long", Some(ErrorCode::ContextLengthExceeded), Some("context_length_exceeded")),
        (json!({ "type": "response.failed" }), "Codex response failed", None, None),
        (json!({ "type": "response.incomplete", "response": { "incomplete_details": { "reason": "max_output_tokens" } } }), "Incomplete Codex response returned, reason: max_output_tokens", Some(ErrorCode::ContextLengthExceeded), None),
        (json!({ "type": "response.incomplete", "response": { "incomplete_details": { "reason": "content_filter" } } }), "Incomplete Codex response returned, reason: content_filter", Some(ErrorCode::Incomplete), None),
        (json!({ "type": "error", "code": "server_error", "message": "backend failed" }), "backend failed", Some(ErrorCode::Overloaded), Some("server_error")),
        (json!({ "type": "error", "error": { "type": "invalid_request_error", "message": "Invalid prompt_cache_key" }, "status": 400 }), "Invalid prompt_cache_key", Some(ErrorCode::Vendor("invalid_request_error".into())), Some("invalid_request_error")),
        // A code that is not a string reads as absent, so the message still
        // reaches the user.
        (json!({ "type": "error", "message": "upstream failed", "code": 500 }), "upstream failed", None, None),
        (json!({ "type": "error" }), "Codex stream error", None, None),
    ];
    for (frame, message, code, provider_code) in cases {
        let events = events(&[frame.clone(), json!({ "type": "response.output_text.delta", "delta": "never read" })]).await;
        let failure = failure_of(events);
        assert_eq!((failure.message.as_str(), &failure.code), (message, &code), "{frame}");
        let diagnostics = failure.diagnostics.unwrap();
        assert_eq!(diagnostics.provider_code.as_deref(), provider_code, "{frame}");
        let upstream: Value = serde_json::from_str(diagnostics.upstream.as_deref().unwrap()).unwrap();
        assert_eq!(upstream, frame);
    }
}

#[tokio::test]
async fn a_failure_keeps_the_request_and_response_ids_it_carries() {
    let failure = failure_of(
        events(&[json!({ "type": "response.failed", "response": { "id": "resp-1", "error": { "code": "server_error", "message": "Failed. Please include the request ID req-1 in your message." } } })]).await,
    );
    assert_eq!(failure.code, Some(ErrorCode::Overloaded));
    let ProviderErrorDiagnostics { provider_request_id, provider_response_id, .. } = failure.diagnostics.unwrap();
    assert_eq!((provider_request_id.as_deref(), provider_response_id.as_deref()), (Some("req-1"), Some("resp-1")));
    let named = failure_of(events(&[json!({ "type": "error", "error": { "message": "failed", "request_id": "req-2" } })]).await);
    assert_eq!(named.diagnostics.unwrap().provider_request_id.as_deref(), Some("req-2"));
}

#[tokio::test]
async fn a_malformed_mapped_event_is_a_protocol_failure_that_names_its_field() {
    let cases = [
        (json!({ "type": "response.output_item.done", "item": { "type": "message", "content": 42 } }), "content"),
        (json!({ "type": "response.output_text.delta", "delta": 42 }), "delta"),
        (json!({ "type": "response.completed", "response": { "usage": { "input_tokens": "100" } } }), "input_tokens"),
        (json!({ "type": "response.completed", "response": { "usage": { "input_tokens": 12.5 } } }), "input_tokens"),
        (json!({ "type": "response.output_item.done", "item": { "type": "reasoning", "summary": [{ "type": "summary_text" }] } }), "text"),
    ];
    for (frame, field) in cases {
        let failure = failure_of(events(std::slice::from_ref(&frame)).await);
        assert_eq!(failure.code, None, "{frame}");
        assert!(failure.message.contains(field), "{frame}: {}", failure.message);
        let diagnostics = failure.diagnostics.unwrap();
        assert_eq!(diagnostics.source, FailureSource::Stream);
        assert_eq!(diagnostics.upstream, Some(frame.to_string()));
    }
    // An event or item type Demi does not map decodes to nothing.
    assert_eq!(decode_frame(r#"{"type":"response.queued","id":"r1"}"#).unwrap(), None);
    assert_eq!(decode_frame(" [DONE] ").unwrap(), None);
    assert!(decode_frame(r#"{"delta":"no type"}"#).is_err());
}

#[tokio::test]
async fn a_cancelled_run_ends_without_a_further_event() {
    let cancel = CancellationToken::new();
    cancel.cancel();
    let frames = body(&[json!({ "type": "response.output_text.delta", "delta": "hi" })]);
    let events: Vec<ProviderEvent> = map_events(sse_events(frames, "Codex", cancel.clone()), vendor(read_http_failure), "codex:", cancel)
        .collect()
        .await;
    assert!(events.is_empty(), "{events:?}");
}
